client/verta/verta/_internal_utils/_histogram_utils.py

```diff
@@ -0,0 +1,209 @@
+            except ValueError:
+                pass
+
+        if isinstance(value, (six.integer_types, float)):
+            if value == 0:
+                zeros += 1
+                continue
+            elif value == 1:
+                ones += 1
+                continue
+
+        # unsupported value
+        raise HistogramError("invalid binary value {}".format(value))
+
+    return {
+        'histogram': {
+            'binary': {
+                'count': [zeros, ones],
+            },
+        },
+        'type': "binary",
+    }
+
+def calculate_discrete_histogram(data):
+    """
+    Calculates a histogram for discrete `data`.
+
+    Parameters
+    ----------
+    data : pandas.Series of int
+        Discrete data to be binned.
+
+    Returns
+    -------
+    histogram : dict
+
+    Raises
+    ------
+    HistogramError
+        If a discrete histogram cannot be calculated from `data`.
+
+    """
+    value_counts = data.value_counts().sort_index()
+    values = value_counts.index.tolist()
+    counts = value_counts.values.tolist()
+
+    # reject non-numbers
+    try:
+        values = list(map(float, values))
+    except ValueError:
+        raise HistogramError(
+            "values must be castable to numbers"
+        )
+
+    # reject non-integral floats
+    if not all(value.is_integer() for value in values):
+        raise HistogramError(
+            "values must be integers"
+        )
+    values = list(map(int, values))
+
+    # heuristic: reject if too many values
+    if len(values) > 10:
+        raise HistogramError(
+            "got {} possible discrete values but heuristic says the maximum is 10".format(len(values))
+        )
+
+    # heuristic: reject if counts don't seem high enough
+    if value_counts.mean() < 10:  # `value_counts` instead of `counts` for mean() method
+        raise HistogramError(
+            "heuristic says that each discrete value should average at least 10 appearances"
+        )
+
+    return {
+        'histogram': {
+            'discrete': {
+                'bucket_values': values,
+                'count': counts,
+            },
+        },
+        'type': "discrete",
+    }
+
+def calculate_float_histogram(data, num_bins=10):
+    """
+    Calculates a histogram for continuous `data`.
+
+    Parameters
+    ----------
+    data : pandas.Series of float
+        Continuous data to be binned.
+    num_bins : int, default 10
+        Number of bins to use.
+
+    Returns
+    -------
+    histogram : dict
+
+    """
+    values = data.values.tolist()
+
+    # reject non-numbers
+    try:
+        values = list(map(float, values))
+    except ValueError:
+        raise TypeError(
+            "unable to generate histogram from non-numeric column {}".format(data.name)
+        )
+
+    # calculate bin boundaries
+    start, stop = min(values), max(values)
+    space = (stop - start)/num_bins
+    bin_boundaries = [start + space*i for i in range(num_bins+1)]
```

Comment:
Please make sure that the last bin boundary is equal to the stop value. It will help avoid numerical issues.